api/tests/test_database.py

```diff
@@ -83,18 +85,100 @@ def test_merge_gtfs_feed(test_database):
+
+
+def test_generate_unique_id():
+    unique_id = generate_unique_id()
+    assert len(unique_id) == 36  # UUIDs are 36 characters long
+    generated_ids = []
+    for _ in range(100):
+        generated_ids.append(generate_unique_id())
+    unique_ids = set(generated_ids)
+    assert len(unique_ids) == len(generated_ids)  # All ids are unique
+
+
+def test_database_connection():
+    db = Database()
+    assert db.is_connected()
+
+
+def test_insert_and_select():
+    db = Database()
+    component_name = fake.name()
+    new_component = Component(name=component_name)
+    db.merge(new_component, auto_commit=True)
+    retrieved_components = db.select(Component, conditions=[Component.name == component_name])
+    assert len(retrieved_components) == 1
+    assert retrieved_components[0][0].name == component_name
+
+    # Ensure no active session exists
+    if db.session:
+        db.close_session()
+
+    results_after_session_closed = db.select_from_active_session(Component)
+    assert len(results_after_session_closed) == 0
+
+
+def test_select_from_active_session_success():
+    db = Database()
+
+    component_name = fake.name()
+    new_component = Component(name=component_name)
+    db.session.add(new_component)
+
+    # The active session should have one instance of the component
+    conditions = [Component.name == component_name]
+    selected_components = db.select_from_active_session(Component, conditions=conditions, attributes=["name"])
+    all_components = db.select(Component)
+    assert len(all_components) >= 1
+    assert len(selected_components) == 1
+    assert selected_components[0]["name"] == component_name
+
+    db.session.rollback()
+
+    # The database should have no instance of the component
+    retrieved_components = db.select(Component, conditions=[Component.name == component_name])
+    assert len(retrieved_components) == 0
+
+
+def test_merge_relationship_w_uncommitted_changed():
+    db = Database()
+    db.start_session()
+
+    # Create and add a new Component object (parent) to the session
+    component_name = fake.name()
+    new_component = Component(name=component_name)
+    db.merge(new_component)
+
+    # Create a new GtfsDataset object (child)
+    gtfs_dataset_id = fake.uuid4()
+    new_gtfs_dataset = Gtfsdataset(id=gtfs_dataset_id)
+
+    # Merge this GtfsDataset into the Component's datasets relationship
+    db.merge_relationship(
+        parent_model=Component,
+        parent_key_values={"name": component_name},
+        child=new_gtfs_dataset,
+        relationship_name="datasets",
+        auto_commit=False,
+        uncommitted=True,
+    )
+
+    # Retrieve the component and check if the GtfsDataset was added
+    retrieved_component = db.select_from_active_session(Component, conditions=[Component.name == component_name])[0]
+    dataset_ids = [dataset.id for dataset in retrieved_component.datasets]
+    assert gtfs_dataset_id in dataset_ids
+
+    # Clean up
+    db.session.rollback()
```

Comment:
[suggestion]: add a try/catch statement with a finally and add the rollback to the finally section. This prevents leaving rows behind in case there is an error raised by the test

Reply:
Done ✅